src/libsyntax/parse/parser.rs

```diff
@@ -7199,44 +7199,41 @@ impl<'a> Parser<'a> {
             return Ok(Some(item));
         }
 
-        // `unsafe async fn` or `async fn`
-        if (
-            self.check_keyword(kw::Unsafe) &&
-            self.look_ahead(1, |t| t.is_keyword(kw::Async))
-        ) || (
-            self.check_keyword(kw::Async) &&
-            self.look_ahead(1, |t| t.is_keyword(kw::Fn))
-        )
-        {
-            // ASYNC FUNCTION ITEM
-            let unsafety = self.parse_unsafety();
-            self.expect_keyword(kw::Async)?;
-            let async_span = self.prev_span;
-            self.expect_keyword(kw::Fn)?;
-            let fn_span = self.prev_span;
-            let (ident, item_, extra_attrs) =
-                self.parse_item_fn(unsafety,
-                                   respan(async_span, IsAsync::Async {
-                                       closure_id: ast::DUMMY_NODE_ID,
-                                       return_impl_trait_id: ast::DUMMY_NODE_ID,
-                                       arguments: Vec::new(),
-                                   }),
-                                   respan(fn_span, Constness::NotConst),
-                                   Abi::Rust)?;
-            let prev_span = self.prev_span;
-            let item = self.mk_item(lo.to(prev_span),
-                                    ident,
-                                    item_,
-                                    visibility,
-                                    maybe_append(attrs, extra_attrs));
-            if self.span.rust_2015() {
-                self.diagnostic().struct_span_err_with_code(
-                    async_span,
-                    "`async fn` is not permitted in the 2015 edition",
-                    DiagnosticId::Error("E0670".into())
-                ).emit();
+        // Parse `async unsafe? fn`.
+        if self.check_keyword(kw::Async) {
+            let async_span = self.span;
+            if self.look_ahead(1, |t| t.is_keyword(kw::Fn))
+                || self.look_ahead(2, |t| t.is_keyword(kw::Fn))
+            {
+                // ASYNC FUNCTION ITEM
+                self.bump(); // `async`
```

Comment:
nit: Why `bump` rather than `expect_keyword`? (same below)

Personally I think the expect is a bit nicer since it doesn't require the extra documentation and provides an assert for the future in case things get changed / moved around.

Reply:
In this case (not the one below), doesn't this amount to doing:
```rust
self.expected_tokens.push(TokenType::Keyword(kw));
```
twice?

(that's the reason I did `.bump()` here)

Reply:
because you already called `check_keyword`? yeah, but IMO that's far enough above as to make it non-obvious to someone who might consider inserting code in the middle, but it doesn't matter too much either way

Reply:
> because you already called `check_keyword`?

Yeah; I wonder if it would be incorrect to do it twice basically...

The second `.bump()` should be fine to change at least if you prefer that.

Reply:
Up to you.

Reply:
Replaced the second `.bump()` with `self.expect_keyword(kw::Fn)?;`. :)